html/infrastructure/safe-passing-of-structured-data/transfer-errors.window.js

```diff
@@ -0,0 +1,42 @@
+function assert_transfer_error(transferList, exception = "DataCloneError") {
+  assert_throws(exception, () => self.postMessage({ get whatever() { throw exception } }, "*", transferList));
+}
+
+test(() => {
+  const b = new ArrayBuffer(1);
+  assert_transfer_error([b, b]);
+}, "Cannot transfer the same object twice");
+
+test(() => {
+  [self, self.document, new Image()].forEach(val => {
+    assert_transfer_error([val]);
+  });
+}, "Cannot transfer all objects");
+
+test(() => {
+  const b = new ArrayBuffer(1);
+  self.postMessage(null, "*", [b]);
+  assert_transfer_error([b], new Error("hi"));
+  assert_throws("DataCloneError", () => self.postMessage(null, '*', [b]));
+}, "Serialize throws before a transferred detached ArrayBuffer is found");
+
+promise_test(() => {
+  return self.createImageBitmap(document.createElement("canvas")).then(bitmap => {
```

Comment:
Do we want to be more exhaustive? We could do something like:

```js
function transfer_tests(name, create) {
  promise_test(async () => {
    const transferable = await create();
    assert_transfer_error([transferable, transferable]);
  }, `Cannot transfer the same ${name} twice`);

  promise_test(async () => {
    const transferable = await create();
    self.postMessage(null, "*", [transferable]);
    assert_throws("DataCloneError", () => self.postMessage(null, "*", [transferable]);
  }, `Serialize should make the ${name} detached, so it cannot be transferred again.`);

  promise_test(async () => {
    const transferable = await create();
    self.postMessage(null, "*", [transferable]);
    assert_transfer_error([transferable], new Error("hi"));
  }, `Serialize should throw before a detached ${name} is found.`);

  promise_test(async () => {
    const transferable = await create();
    let seen = false;
    const message = {
      get a() {
        self.postMessage(null, '*', [transferable]);
        seen = true;
      }
    };
    assert_throws("DataCloneError", () => self.postMessage(message, "*", [transferable]));
    assert_true(seen);
  }, `Cannot transfer ${name} detached while the message was serialized.`);
}

transfer_tests("ArrayBuffer", () => new ArrayBuffer(1));
transfer_tests("MessagePort", () => new MessageChannel().port1);
transfer_tests("ImageBitmap", () => self.createImageBitmap(document.createElement("canvas")));
// OffscreenCanvas? I don't believe we have any other transferables, at least implemented in Blink.
```

That might be excessively clever, but I think the logic for all of these tests generalizes to all kinds of transferable checks, so why not check the whole matrix?

Reply:
Yeah, why not. I'll make it so. Probably not today unfortunately.